src/pytools/api/_api.py

```diff
@@ -323,33 +400,58 @@ def validate_element_types(
         )
 
     if expected_type is not object:
-        if optional:
-            expected_type = (
-                (*expected_type, type(None))
-                if isinstance(expected_type, tuple)
-                else (expected_type, type(None))
-            )
+        expected_type = _as_optional_type(type_=expected_type, is_optional=optional)
+
         for element in iterable:
             if not isinstance(element, expected_type):
-                if name:
-                    message_head = f"{name} requires"
-                else:
-                    message_head = "expected"
-
-                if isinstance(expected_type, type):
-                    expected_type_str = expected_type.__name__
-                else:
-                    expected_type_str = (
-                        f"one of {{{', '.join(t.__name__ for t in expected_type)}}}"
-                    )
-                raise TypeError(
-                    f"{message_head} instances of {expected_type_str} "
-                    f"but got a {type(element).__name__}"
+                _raise_type_mismatch(
+                    name=name,
+                    expected_type=expected_type,
+                    mismatched_type=type(element),
+                    is_single=False,
                 )
 
     return iterable
 
 
+def _as_optional_type(
+    type_: Union[type, Tuple[type, ...]], *, is_optional: bool
+) -> Union[type, Tuple[type, ...]]:
+    # if is_optional is True, return a tuple comprising the original (types) and None
+    if is_optional:
+        if isinstance(type_, tuple):
+            return (*type_, type(None))
+        else:
+            return (type_, type(None))
+    else:
+        return type_
+
+
+def _raise_type_mismatch(
+    *,
+    name: Optional[str],
+    expected_type: Union[type, Tuple[type, ...]],
+    mismatched_type: type,
+    is_single: bool,
```

Comment:
I don't fully get why we need `is_single`. By "X expects instance of A" and "X expects instances of B" you indicate that a collection of such instances needs to be passed?

Reply:
Otherwise I don't see why such differentiation should be kept.

Reply:
@mtsokol good catch! It's a remnant of refactoring, and completely redundant.

Reply:
When called from `validate_type`, we are validating the type of a single object.

When called from `validate_element_types`, we are validating the type of multiple objects.

Arg `is_single` adjusts the wording of the error message depending on this difference.

Reply:
Then all clear to me!